File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/js/dp-10-longest-common-subsequence.js
/*
Title: Longest Common Subsequence
Problem: Given two strings, return the length of their longest common subsequence (not necessarily contiguous).
Constraints: 1 <= length <= 15, lowercase letters
Examples:
- Input: "abcde" and "ace" => Output: 3
- Input: "abc" and "abc" => Output: 3
- Input: "abc" and "def" => Output: 0
Edge cases: No common characters.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

function solve(text1, text2) {
  // TODO: implement
  return 0;
}


function refSolve(a, b) {
  const m = a.length;
  const n = b.length;
  const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));
  for (let i = 1; i <= m; i += 1) {
    for (let j = 1; j <= n; j += 1) {
      if (a[i - 1] === b[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
      else dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
    }
  }
  return dp[m][n];
}


function assertEqual(actual, expected, message) {
  if (actual !== expected) {
    throw new Error(`Fail: ${message}. Expected ${expected}, got ${actual}`);
  }
}

function assertDeepEqual(actual, expected, message) {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) {
    throw new Error(`Fail: ${message}. Expected ${b}, got ${a}`);
  }
}

let seed = 123456789;
function rand() {
  seed = (seed * 1664525 + 1013904223) >>> 0;
  return seed;
}

function randInt(min, max) {
  return min + (rand() % (max - min + 1));
}

let passed = 0;
let failed = 0;

function run(name, fn) {
  try {
    fn();
    passed += 1;
  } catch (err) {
    failed += 1;
    console.error(`Test failed: ${name}`);
    console.error(err.message);
  }
}

run('deterministic', () => {

  assertEqual(solve("abcde", "ace"), 3, 'deterministic');
  assertEqual(solve("abc", "abc"), 3, 'deterministic');
  assertEqual(solve("abc", "def"), 0, 'deterministic');
  assertEqual(solve("aggtab", "gxtxayb"), 4, 'deterministic');
  assertEqual(solve("aaaa", "aa"), 2, 'deterministic');
  assertEqual(solve("ab", "ba"), 1, 'deterministic');

});

run('random', () => {

  const letters = ['a', 'b', 'c'];
for (let t = 0; t < 20; t += 1) {
  const len1 = randInt(1, 6);
  const len2 = randInt(1, 6);
  const a = Array.from({ length: len1 }, () => letters[randInt(0, letters.length - 1)]).join('');
  const b = Array.from({ length: len2 }, () => letters[randInt(0, letters.length - 1)]).join('');
  const expected = refSolve(a, b);
  assertEqual(solve(a, b), expected, `a=${a}, b=${b}`);
}

});

console.log(`Passed: ${passed}, Failed: ${failed}`);

if (failed > 0) process.exit(1);

module.exports = { solve };

File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/c99/dp-10-longest-common-subsequence.c
/*
Title: Longest Common Subsequence
Problem: Given two strings, return the length of their longest common subsequence (not necessarily contiguous).
Constraints: 1 <= length <= 15, lowercase letters
Examples:
- Input: "abcde" and "ace" => Output: 3
- Input: "abc" and "abc" => Output: 3
- Input: "abc" and "def" => Output: 0
Edge cases: No common characters.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int passed = 0;
static int failed = 0;

#define ASSERT_INT_EQ(actual, expected, msg) do { \
  if ((actual) != (expected)) { \
    failed += 1; \
    printf("FAIL: %s (expected %d, got %d)\\n", msg, (expected), (actual)); \
  } else { \
    passed += 1; \
  } \
} while (0)

#define ASSERT_BOOL_EQ(actual, expected, msg) ASSERT_INT_EQ((actual) != 0, (expected) != 0, msg)

static unsigned int seed = 123456789u;
static unsigned int lcg(void) {
  seed = seed * 1664525u + 1013904223u;
  return seed;
}

static int rand_int(int min, int max) {
  return min + (int)(lcg() % (unsigned int)(max - min + 1));
}


int solve(const char *a, const char *b) {
  // TODO: implement
  return 0;
}


static int ref_solve(const char *a, const char *b) {
  int m = (int)strlen(a);
  int n = (int)strlen(b);
  int *dp = (int *)calloc((m + 1) * (n + 1), sizeof(int));
  for (int i = 1; i <= m; i += 1) {
    for (int j = 1; j <= n; j += 1) {
      int idx = i * (n + 1) + j;
      if (a[i - 1] == b[j - 1]) dp[idx] = dp[(i - 1) * (n + 1) + j - 1] + 1;
      else {
        int up = dp[(i - 1) * (n + 1) + j];
        int left = dp[i * (n + 1) + j - 1];
        dp[idx] = up > left ? up : left;
      }
    }
  }
  int ans = dp[m * (n + 1) + n];
  free(dp);
  return ans;
}


static void test_deterministic(void) {

  ASSERT_INT_EQ(solve("abcde", "ace"), 3, "deterministic");
  ASSERT_INT_EQ(solve("abc", "abc"), 3, "deterministic");
  ASSERT_INT_EQ(solve("abc", "def"), 0, "deterministic");
  ASSERT_INT_EQ(solve("aggtab", "gxtxayb"), 4, "deterministic");
  ASSERT_INT_EQ(solve("aaaa", "aa"), 2, "deterministic");
  ASSERT_INT_EQ(solve("ab", "ba"), 1, "deterministic");

}

static void test_random(void) {

  const char letters[] = "abc";
for (int t = 0; t < 20; t += 1) {
  int len1 = rand_int(1, 6);
  int len2 = rand_int(1, 6);
  char a[8];
  char b[8];
  for (int i = 0; i < len1; i += 1) a[i] = letters[rand_int(0, 2)];
  a[len1] = '\0';
  for (int j = 0; j < len2; j += 1) b[j] = letters[rand_int(0, 2)];
  b[len2] = '\0';
  int expected = ref_solve(a, b);
  ASSERT_INT_EQ(solve(a, b), expected, "random lcs");
}

}

int main(void) {

  test_deterministic();

  test_random();

  printf("Passed: %d, Failed: %d\n", passed, failed);

  return failed > 0 ? 1 : 0;

}

